sematic/plugins/building/docker.py

```diff
@@ -0,0 +1,472 @@
+
+        # TODO: configure the docker service connection string in the build file
+        docker_client = docker.from_env()  # type: ignore
+
+        # TODO: pull if it is not available locally
+        image = docker_client.images.get(str(base_image_uri))
+        if image.id != base_image_uri.digest:
+            raise BuildError(
+                f"Digest mismatch for image '{base_image_uri}'; "
+                f"expected: {base_image_uri.digest} got: {image.id}"
+            )
+
+        logger.info("Found base image: '%s'", base_image_uri)
+
+        build_image_uri = _push_image(
+            image=image,
+            image_uri=base_image_uri,
+            push_config=build_config.push,
+            docker_client=docker_client,
+        )
+
+        logger.info("Finished building image: '%s'", repr(build_image_uri))
+
+        return build_image_uri
+
+    def _launch(self, target: str, image_uri: ImageURI) -> None:
+        """
+        Launches the specified user code target, using the specified image.
+        """
+        sys.path.append(os.getcwd())
+        # TODO: revert this overwrite after finishing execution
+        #  promote the `environment_variables` testing fixture to a utility
+        os.environ[CONTAINER_IMAGE_ENV_VAR] = repr(image_uri)
+
+        logger.info("Launching target: '%s'", target)
+
+        runpy.run_path(path_name=target, run_name="__main__")
+
+        logger.info("Finished launching target: '%s'", target)
+
+
+def _get_build_config(script_path: str) -> BuildConfig:
+    """
+    Handles the entire instantiation of the definite `BuildConfig` object for the
+    specified script.
+
+    Raises
+    ------
+    BuildConfigurationError:
+        There was an error when loading or validating the specified build configuration.
+    """
+    build_config_files = _find_build_config_files(script_path=script_path)
+    logger.info(
+        "Script '%s' has these corresponding build files: %s",
+        script_path,
+        build_config_files,
+    )
+
+    # TODO: override several config files, env vars, and cli arguments
+    if len(build_config_files) == 0:
+        raise BuildConfigurationError(
+            f"Unable to find any build files corresponding to script '{script_path}'! "
+            f"Please see TODO for build configuration details!"
+        )
+
+    return BuildConfig.load_build_config_file(build_config_files[0])
+
+
+def _find_build_config_files(script_path: str) -> List[str]:
+    """
+    Searches for build configuration files that correspond to the specified script file,
+    and returns a list containing their paths.
+    """
+    # TODO: also load `build.yaml` files on the file hierarchy between the cwd and the
+    #  script directory
+    root = os.path.splitext(script_path)[0]
+    file_candidate = f"{root}.yaml"
+
+    return [file_candidate] if os.path.isfile(file_candidate) else []
+
+
+def _get_image_uri(build_config: BuildConfig, target: str) -> ImageURI:
```

Comment:
This function should be named differently since it actually makes potentially heavy system calls. For example `_run_build_script`.